src/gennylib/src/topology.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <optional>
+#include <bsoncxx/json.hpp>
+#include <bsoncxx/types.hpp>
+#include <boost/log/trivial.hpp>
+
+#include <gennylib/topology.hpp>
+
+namespace genny {
+
+std::string Topology::nameToUri(const std::string& name) {
+    std::set<std::string> nameSet;
+    // Hostnames returned from some commands have the form configRepl/hostname:port
+    auto strippedName = name.substr(name.find('/') + 1, std::string::npos);
+    nameSet.insert(strippedName);
+    _factory.overrideHosts(nameSet);
+    return _factory.makeUri();
+}
+
+void Topology::getDataMemberConnectionStrings(DBService& service) {
+    auto res = service.runAdminCommand("isMaster");
+    if (!res.view()["setName"]) {
+        std::unique_ptr<MongodDescription> desc = std::make_unique<MongodDescription>();
```

Comment:
maybe make the uri a ctor arg of the `MongoDescription`? (there are a few other things here that maybe want to be pushed into impl details (perhaps in ctors) of the `*Description` classes. Not sure of implications on testing but may actually make it easier so not a strong opinion here yet.